user-console/src/main/java/org/pentaho/mantle/client/admin/EmailAdminPanelController.java

```diff
@@ -178,6 +196,84 @@ public void onKeyUp( final KeyUpEvent keyUpEvent ) {
         setDirty( true );
       }
     } );
+
+    prepareTextBox( clientIdTextBox, new ChangeHandler() {
+      @Override
+      public void onChange( ChangeEvent event ) {
+        emailConfig.setClientId( clientIdTextBox.getValue() );
+        setDirty( true );
+      }
+    } );
+
+    prepareTextBox( clientSecretTextBox, new ChangeHandler() {
+      @Override
+      public void onChange( ChangeEvent event ) {
+        emailConfig.setClientSecret( clientSecretTextBox.getValue() );
+        setDirty( true );
+      }
+    } );
+
+    prepareTextBox( scopeTextBox, new ChangeHandler() {
+      @Override
+      public void onChange( ChangeEvent event ) {
+        emailConfig.setScope( scopeTextBox.getValue() );
+        setDirty( true );
+      }
+    } );
+
+    prepareTextBox( tokenUrlTextBox, new ChangeHandler() {
+      @Override
+      public void onChange( ChangeEvent event ) {
+        emailConfig.setTokenUrl( tokenUrlTextBox.getValue() );
+        setDirty( true );
+      }
+    } );
+
+    prepareTextBox( authorizationCodeTextBox, new ChangeHandler() {
+      @Override
+      public void onChange( ChangeEvent event ) {
+        emailConfig.setAuthorizationCode( authorizationCodeTextBox.getValue() );
+        setDirty( true );
+      }
+    } );
+
+    prepareTextBox( refreshTokenTextBox, new ChangeHandler() {
+      @Override
+      public void onChange( ChangeEvent event ) {
+        emailConfig.setRefreshToken( refreshTokenTextBox.getValue() );
+        setDirty( true );
+      }
+    } );
+
+    prepareTextBox( redirectUriTextBox, new ChangeHandler() {
+      @Override
+      public void onChange( ChangeEvent event ) {
+        emailConfig.setRedirectUri( redirectUriTextBox.getValue() );
+        setDirty( true );
+      }
+    } );
+
+    authenticationTypeLB.addChangeHandler( new ChangeHandler() {
+      public void onChange( final ChangeEvent changeEvent ) {
+        String authMechanism = authenticationTypeLB.getValue( authenticationTypeLB.getSelectedIndex() );
+        oauthPanel.setVisible( authMechanism.equalsIgnoreCase( "XOAUTH2" ) );
+        passwordPanel.setVisible( authMechanism.equalsIgnoreCase( "basic" ) );
```

Comment:
These constants everywhere don't look right to me.

I propose you define a class to hold the constants of each kind, or just use an existing interface to hold these values.

For example, you could add constants to the `IEmailConfiguration` interface:

```java
public interface IEmailConfiguration {
  // Authentication Types
  public static final String AUTH_TYPE_XOAUTH2 = "XOAUTH2";
  public static final String AUTH_TYPE_BASIC = "basic";

  // Grant Types
  public static final String GRANT_TYPE_REFRESH_TOKEN = "refresh_token";
  public static final String GRANT_TYPE_AUTH_CODE = "authorization_code";
  
  // Protocols
  public static final String PROTOCOL_GRAPH_API = "graph_api";
  // ...

  // ...Any Other Known Values Of Specific Kinds of Things...
}
```

Also, please settle on the value being **either** upper or lower case and compare it always in a case sensitive manner (using `.equals( . )`. The way it is, there is unnecessary ambiguity and inconsistency. If needed, _normalize_ values received from uncontrolled origins — or just fail if value not provided verbatim as needed.

Use these constants from any of the files needing them, instead of declaring local constants. You can even use static imports to pretend they are local constants.

Reply:
I concur create enums or java constants

Reply:
This is still an open issue.
See https://github.com/pentaho/pentaho-platform/pull/5185#discussion_r1086001880 and other unresolved comments.

Reply:
Apparently, in a comment that I cannot see, you already created an `EmailConstants` class.